Nested Keys `payload_selector`

Doordashcon opened the pull request:
Addresses https://github.com/qdrant/qdrant/issues/2336

Enables nested keys to extract payload properties

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

agourlay commented:
Looking forward to see tests for this change :+1:

Doordashcon commented:
It seems without any change made all python integration tests are failing.

Doordashcon commented:
Hello @agourlay Thanks for the insight

This somewhat relates with my question here?
https://github.com/qdrant/qdrant/issues/2336#issuecomment-1722957877

agourlay commented:
I do not think the issue here is related to your question.

Regarding that question, I agree with Luis about preserving the path to the target.

Also, your change must be backward compatible with the current API. 

Meaning, the selector becomes a string `JsonPath` pointing to a JSON key which could be at an arbitrary depth level, including the root level.

In the following code you are unpacking the target of the selector, but you forget to do it in case of scalar value:

```rust
     for value in result {
            match value {
                Value::Array(inner) => {
                    for inner_value in inner {
                        if let Value::Object(ref obj) = inner_value {
                            map.extend(obj.clone());
                        }
                    }
                }
                Value::Object(ref obj) => {
                    map.extend(obj.clone());
                }
                _ => {} // Here must handle scalar as well
           
```

The grouping API uses a root selector on the "docId" field which returns an id.

agourlay commented:
@Doordashcon Have you tried reusing our JsonPath infrastructure in https://github.com/qdrant/qdrant/blob/dev/lib/segment/src/common/utils.rs#L204 ?

It could do most of the heavy lifting for you :)

Doordashcon commented:
using the `get_value` method employs
```rust
impl PayloadContainer for Payload {
    fn get_value(&self, path: &str) -> MultiValue<&Value> {
        utils::get_value_from_json_map(path, &self.0)
    }
}```

Doordashcon commented:
Also been having some issues running the tests present in the workflow file locally.

This error originates from the first workflow test
```bash
raise ConnectionError(e, request=request)
E           requests.exceptions.ConnectionError: HTTPConnectionPool(host='localhost', port=6333): Max retries exceeded with url:

/collections/test_collection/points/search (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0xffff81414650>: 

Failed to establish a new connection: [Errno 111] Connection refused'))
```

agourlay commented:
The `openapi` integration tests require that you start manually a local qdrant instance.
Is this maybe what you are missing?

agourlay commented:
Also you have a failing unit test https://github.com/qdrant/qdrant/actions/runs/6312187511/job/17178790848?pr=2669#step:8:485

```
--- TRY 3 STDERR:        collection::integration collection_restore_test::test_collection_payload_custom_payload ---
thread 'collection_restore_test::test_collection_payload_custom_payload' panicked at 'index out of bounds: the len is 0 but the index is 0', lib/segment/src/types.rs:1736:43
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Doordashcon commented:
> Also you have a failing unit test https://github.com/qdrant/qdrant/actions/runs/6312187511/job/17178790848?pr=2669#step:8:485
> 
> ```
> --- TRY 3 STDERR:        collection::integration collection_restore_test::test_collection_payload_custom_payload ---
> thread 'collection_restore_test::test_collection_payload_custom_payload' panicked at 'index out of bounds: the len is 0 but the index is 0', lib/segment/src/types.rs:1736:43
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ```

Currently writing more tests to see to this

Doordashcon commented:
> The `openapi` integration tests require that you start manually a local qdrant instance. Is this maybe what you are missing?

current testing workflow
- build: `cargo build --features service_debug`
- run `cargo run`
- test `./tests/integration-tests.sh`

Doordashcon commented:
> Also been having some issues running the tests present in the workflow file locally.
> 
> This error originates from the first workflow test
> 
> ```shell
> raise ConnectionError(e, request=request)
> E           requests.exceptions.ConnectionError: HTTPConnectionPool(host='localhost', port=6333): Max retries exceeded with url:
> 
> /collections/test_collection/points/search (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0xffff81414650>: 
> 
> Failed to establish a new connection: [Errno 111] Connection refused'))
> ```

Still encountering the above error message, after which all test fail. I made sure to clear all cache and kill running networks before using the `./tests/integration-tests.sh` command on it's own.

agourlay commented:
Is your qdrant instance actually up and listening to the port 6333?
Check your logs.

agourlay commented:
@Doordashcon I have decided to take over this feature given that you have problem with your local setup and that the feature is more involved than I initially thought.

Hopefully there will be new opportunities for contributions which are more amenable. 

Feel free to review it https://github.com/qdrant/qdrant/pull/2756

Doordashcon commented:
Hi @agourlay, my apologies for the delayed response, in search of a linux setup currently. apparently docker on Mac has a few hiccups.


Added some modification to the `process` method, which clears all unit tests, also did extensive testing using the qdrant python client and all looks good.

```rust
pub fn process(&self, x: Payload) -> Payload {
        let fields = match self {
            PayloadSelector::Include(selector) => selector.include.clone(),
            PayloadSelector::Exclude(selector) => {
                x.0.keys()
                    .filter(|&k| !selector.exclude.contains(k))
                    .cloned()
                    .collect()
            }
        };

        let mut map = Map::new();

        for field in &fields {
            let values = x.get_value(field).values();

            if field.contains('.') {
                let parts: Vec<String> = field.split('.').map(|part| part.to_string()).collect();
                let last_char = parts.last().cloned().unwrap_or_else(|| String::from(""));
                for value in values {
                    map.insert(last_char.clone(), value.clone());
                }
            } else {
                for value in values {
                    map.insert(field.clone(), value.clone());
                }
            }
        }
        map.into()
    }
```
